distributed/worker.py

```diff
@@ -1671,15 +1727,21 @@ def transition_new_waiting(self, ts):
                 pdb.set_trace()
             raise
 
-    def transition_new_fetch(self, ts):
+    def transition_new_fetch(self, ts, who_has):
         try:
             if self.validate:
                 assert ts.state == "new"
                 assert ts.runspec is None
+                # assert who_has
```

Comment:
Right, I had some code in `add_task` at some point which removes `self.address` from who_has in case the task was not im memory. I should check this again. This is important.

Reply:
The assert is back in. This edge case is exclusively handled in update_who_has and we will emit a log if this is encountered. At least I couldn't reproduce any wrongful state anymore